### ISSUE TITLE
Improve Bikram Sambat date picker

### ISSUE THREAD

**Issue by binokaryg** [medic/cht-core#9250]
**What feature do you want to improve?**
The [Bikram Sambat date picker](https://docs.communityhealthtoolkit.org/bikram-sambat/) we currently have for [the forms](https://docs.communityhealthtoolkit.org/apps/reference/forms/app/#bikram-sambat-datepicker) provides an easy way to enter dates in Bikram Sambat calendar. However, it lacks many features of a complete date picker.

**Describe the improvement you'd like**

Some of the features missing in the Bikram Sambat date picker compared to its Gregorian counterpart:
- Proper calendar view (Month and Weeks)
- Option to jump to Today
- Option to clear
- Option to go to next and previous month
- Option to select the year and date

**Describe alternatives you've considered**
None

**Additional context**
Screenshot from the [CHT docs](https://docs.communityhealthtoolkit.org/bikram-sambat/):
![image](https://github.com/medic/cht-core/assets/20041732/373d3e8d-b85f-4f2c-9177-c15d8dab5328)

A public repo that has better date picker: https://github.com/opensource-nepal/nepali-date-picker
Demo: https://leapfrogtechnology.github.io/nepali-date-picker/demo/

**Comment by binokaryg** [comment]
Here are some screenshots for reference:

1. [Current](https://docs.communityhealthtoolkit.org/bikram-sambat/) Bikram Sambat date picker:
![image](https://github.com/user-attachments/assets/fb395830-2429-4005-8d48-c3721eff4b16)

2. [Demo](https://leapfrogtechnology.github.io/nepali-date-picker/demo/) Bikram Sambat date picker:

![image](https://github.com/user-attachments/assets/7a67dcd8-b137-4f8a-aeaa-be6e962b7915)

![image](https://github.com/user-attachments/assets/f4cac5c5-3c33-44ec-8cf9-6dd8b7777756)

![image](https://github.com/user-attachments/assets/b239e09a-5a12-4215-a795-c89a93cf2073)

**Comment by binokaryg** [comment]
> [@binokaryg](https://github.com/binokaryg) i would like to work on this issue. can you help for initial setup of cht-core.

Sure, please follow [this guide](https://docs.communityhealthtoolkit.org/contribute/code/core/dev-environment/), and let us know if you need help.

You can post in the [CHT forum](https://forum.communityhealthtoolkit.org/) for help.

**Comment by Akshanshkaushal** [comment]
@binokaryg @sugat009 I would like to work upon this issue. Could you please assign this?

**Comment by Vikaspal8923** [comment]
@andrablaj, I am currently working on another issue, so if @Akshanshkaushal wants to work on this issue, he can take over it. Assign this to @Akshanshkaushal

**Comment by PradyumnChauhan** [comment]
@binokaryg can you tell me how can i access this through frontend from the `localhost:5988/` because im not seeing it in the
` cht-core` frontend

**Comment by jkuester** [comment]
@PradyumnChauhan this issue has already been assigned and is not currently available to work on.  Other good first issues will [be listed here](https://github.com/search?q=is%3Aissue+is%3Aopen+org%3Amedic+label%3A%22Good+First+Issue%22+no%3Aassignee&type=issues) as they become available.

**Comment by mb-oli** [comment]
Its very mandatory, please update like this.

**Comment by Akshanshkaushal** [comment]
@andrablaj @jkuester I have made a PR regarding the implementation; kindly take a look at it.

**Comment by aniket866** [comment]
I would like to work on this 
@sugat009 
@dianabarsan 
please feel free to assign

**Comment by jkuester** [comment]
@aniket866 https://github.com/medic/cht-core/issues/9699#issuecomment-4314617377

**Comment by andrablaj** [comment]
@megha1807 can you please comment on this issue so we can assign the issue to you? Thanks!

**Comment by megha1807** [comment]
Hello @andrablaj ,
Can you please me this issue?

**Comment by sugat009** [comment]
Hi @megha1807, thanks for the upgraded Bikram Sambat datepicker in #11186, the calendar popup is a big step up. While testing it against the original asks in this issue, a few things came up worth revisiting. Sharing them with the snippets I used locally to confirm they work, so they're easy to drop in. Screenshot at the bottom for reference.

**suggestion: keep the calendar open after selecting a date**
The calendar popup is new in #11186 (there was no popup before), and right now selecting a date closes it, so re-picking means reopening. Keeping it open after selection feels smoother; the user can still dismiss it via the × button, the backdrop, or Esc. Two small changes:

```js
$hiddenDateInput.nepaliDatePicker({
  dateFormat: '%y-%m-%d',
  closeOnDateSelect: false
});
```

and in the `dateSelect` handler, drop the `hideDatePicker()` call so it stays open after writing the fields.

**suggestion: add a Clear option (one of the asks that got missed)**
This issue lists "Option to clear" among the desired features, and it looks like it didn't make it into #11186 (this slipped past me in review too). I added a small eraser button in the header next to the Today (home) icon that empties the day, month, and year fields plus the saved value. It sits naturally beside Today, since both are in-calendar actions.

Because the library rebuilds the header on every navigation, the button needs re-injecting after each redraw (guarded), which the existing MutationObserver in the `show` handler already gives a hook for:

```js
const clearDate = () => {
  $parent.find('input[name="day"]').val('');
  $parent.find('input[name="month"]').val('');
  $parent.find('input[name="year"]').val('');
  $group.find('.month-dropdown button').html('महिना <span class="caret"></span>');
  $hiddenDateInput.val('');
  $realDateInput.val('').trigger('change');
};

const ensureClearButton = (container) => {
  const $today = $(container).find('.today-btn');
  if (!$today.length || $(container).find('.clear-btn').length) {
    return;
  }
  const $clear = $('<a href="#" class="clear-btn" title="मिति हटाउनुहोस्"></a>');
  $clear.on('click', function(e) {
    e.preventDefault();
    e.stopPropagation();
    clearDate();
  });
  $today.after($clear);
};
```

wired into the `show` handler alongside the existing sanitize call:

```js
sanitizeHrefs(container);
ensureClearButton(container);
const observer = new MutationObserver(() => {
  sanitizeHrefs(container);
  ensureClearButton(container);
});
```

and the styling in `bikramsambat.less`, nested under `.nepali-date-picker`:

```less
.clear-btn {
  display: inline-block;
  width: 16px;
  height: 16px;
  margin-left: 6px;
  vertical-align: middle;
  cursor: pointer;
  opacity: 0.6;
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24'%3E%3Cpath d='M16.24 3.56l4.95 4.94c.78.79.78 2.05 0 2.84L12 20.53a4.008 4.008 0 0 1-5.66 0L2.81 17c-.78-.79-.78-2.05 0-2.84l10.6-10.6c.79-.78 2.05-.78 2.83 0M4.22 15.58l3.54 3.53c.78.79 2.04.79 2.83 0l3.53-3.53l-4.95-4.95l-4.95 4.95z'/%3E%3C/svg%3E");
  background-size: contain;
  background-repeat: no-repeat;
  background-position: center;

  &:hover {
    opacity: 1;
  }
}
```

One gotcha worth knowing: `bikramsambat.less` re-implements the library's `.nepali-date-picker .icon` sprite rule, so if the clear button also carries the `icon` class, the PNG sprite overrides this custom icon (equal specificity, later rule wins) and it won't render. Keeping the class to just `clear-btn` (no `icon`) avoids it.

**suggestion (minor): close button placement**
The × close button currently sits very close to the next-month arrow and the year dropdown. Floating it to the top-right corner of the modal gives it a cleaner home without crowding the header controls:

Reference: 

<img width="346" height="254" alt="Image" src="https://github.com/user-attachments/assets/4daff3b8-4bdc-46d5-a1ea-0e003135c4ec" />

```less
.nepali-date-picker {
  overflow: visible !important; /* so the corner-floated button is not clipped */

  .close-btn {
    position: absolute;
    top: -12px;
    right: -12px;
    width: 24px;
    height: 24px;
    display: flex;
    align-items: center;
    justify-content: center;
    background: #fff;
    border: 1px solid #ccc;
    border-radius: 50%;
    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.3);
    font-size: 18px;
    cursor: pointer;
    z-index: 10000;
    line-height: 1;
    padding: 0;
    color: #999;

    &:hover {
      color: #333;
    }
  }
}
```

@binokaryg, as the original requester here, your take would be valuable, especially on whether the calendar should stay open after selection and whether a Clear control was part of what you had in mind.

<img width="399" height="285" alt="Bikram Sambat datepicker: eraser button beside the home icon, close button floated to the corner" src="https://github.com/user-attachments/assets/eb4bcae1-fb1d-44e4-82b9-49ab80abc387" />

**Comment by megha1807** [comment]
Thanks, @sugat009 ! These suggestions make a lot of sense and will make the datepicker experience much cleaner.

Since PR #11186 is already merged and I am currently working on implementing the Reports date filter (#9699) using the shared datepicker utility, I can implement these three updates (keeping the calendar open, adding the eraser/clear button, and floating the close button) directly in this current branch. That way, the changes will apply to both Enketo forms and the Reports filters at the same time.

Let me know if I should go ahead and implement these!

**Comment by binokaryg** [comment]
> @binokaryg, as the original requester here, your take would be valuable, especially on whether the calendar should stay open after selection and whether a Clear control was part of what you had in mind.

I agree that keeping the calendar open (unless closed specifically) will allow users to make corrections more easily.  I don't remember thinking about the clear control earlier, but the demo I had linked had a separate "Clear" button. I think it's important to have a clear control especially for those cases where the date field is optional.

**Comment by sugat009** [comment]
**Status: implemented, keeping open until it reaches `master`.**

All the improvements requested here are done and merged into the DMP feature branch (`10707-...`), across #11186 (the upgraded calendar widget) and #11234 (Clear control + keep-open on selection). Mapping to the asks in this issue:

- ✅ Proper calendar view (month + weeks)
- ✅ Jump to Today
- ✅ Clear option
- ✅ Next / previous month navigation
- ✅ Year and date selection
- ✅ Keep the calendar open after selection (per @binokaryg's note)
- ✅ Clear control, especially for optional date fields (per @binokaryg's note)

I confirmed these on the Enketo form widget in-app.

Leaving this open on purpose: the work is on the feature branch, not yet on `master`. It'll be closed once the feature branch lands on master via #11211. Thanks @megha1807 for the implementation.

**Comment by sugat009** [comment]
Hi @binokaryg, an update from building this under the epic #10707.

The calendar we adopted for the grid UI, `leapfrogtechnology/nepali-date-picker`, works fine as an interface, but it carries its **own** hardcoded Bikram Sambat month-length table, separate from our `bikram-sambat` lib. That table is outdated: for example it renders मंसिर २०८२ with 30 days when the month actually has 29. When you pick such a day, `bikram-sambat` (correctly) rejects it and throws `Invalid day value`, which breaks the picker.

I checked its table against a real 2082/2083 calendar: `bikram-sambat` is right, the plugin is stale. And it's effectively unmaintained (last release June 2024), so there's no upstream fix coming.

Since the UI itself is good and it's ISC licensed, the plan is to **fork it into the `medic/bikram-sambat` repo and bind its date logic to `bikram-sambat`** (delegate the month lengths and BS↔AD conversion to the lib). That gives us a single source of truth, so the crash can't recur and the picker automatically tracks any future data fix you make there.

Does that sound right to you? If so, we'll open a feature issue in `medic/bikram-sambat` for the forked picker and a bug in cht-core for the crash plus the swap.

cc @megha1807, who's implementing this in the epic.

**Comment by binokaryg** [comment]
> Since the UI itself is good and it's ISC licensed, the plan is to fork it into the medic/bikram-sambat repo and bind its date logic to bikram-sambat (delegate the month lengths and BS↔AD conversion to the lib). That gives us a single source of truth, so the crash can't recur and the picker automatically tracks any future data fix you make there.\n\nDoes that sound right to you? If so, we'll open a feature issue in medic/bikram-sambat for the forked picker and a bug in cht-core for the crash plus the swap.

Sounds good.

**Comment by sugat009** [comment]
Thanks @binokaryg. Opened the two tickets:

- medic/bikram-sambat#35 — the forked calendar UI bound to `bikram-sambat` (single source of truth), with the licensing steps noted.
- #11252 — the cht-core `Invalid day value` crash plus the picker swap, blocked by #35 and tracked as a sub-issue of the epic #10707.

cc @megha1807.

**Comment by sugat009** [comment]
@binokaryg, a `bikram-sambat` behavior we ran into while reviewing the BS test PRs, would value your read as maintainer (not blocking anything).

When a timestamp that carries a **time** is converted, the BS day rolls at **12:30 UTC (= 6:15 PM Nepal)** rather than at Nepal midnight. So a timestamp still "today" in Nepal but past ~6:15 PM renders the *next* BS day, and disagrees with the Gregorian rendering of the same instant:

- `29 Jun 2024 12:29 UTC` → `१५ असार २०८१`
- `29 Jun 2024 12:30 UTC` → `१६ असार २०८१`
- `29 Jun 2024, 8 PM Nepal` (14:15 UTC) → Gregorian `29 Jun`, BS `१६ असार` (a day apart)

The boundary traces to `BS_EPOCH_TS` sitting at `1913-04-12 12:30:00 UTC`. Date-only string inputs like `toBik('2024-06-29')` are fine; it only surfaces when a `moment`/`Date` carrying a time is passed in, which is what our display path does.

Could you confirm how you'd class this: **intended** (lib is meant for date-only input and callers should strip the time), a **known/accepted** limitation, an actual **bug**, already **accounted for**, or **already ticketed**? Mainly want to know whether it needs a fix and whether our tests should treat current behavior as expected.
CC: @bhishankc

**Comment by binokaryg** [comment]
> When a timestamp that carries a time is converted, the BS day rolls at 12:30 UTC (= 6:15 PM Nepal) rather than at Nepal midnight.

This was [reported by](https://github.com/medic/config-nssd/issues/946) @bhishankc and a bug report was created:
- https://github.com/medic/cht-core/issues/11241

PR with the fix is ready for review:
- https://github.com/medic/cht-core/pull/11251

@sugat009  Would you be able to review it?

**Comment by sugat009** [comment]
Done through #11186 #11234
